### ISSUE TITLE
Support Kubeflow TrainJob (v2)

### ISSUE THREAD

**Issue by tenzen-y** [kubernetes-sigs/kueue#3884]
**What would you like to be added**:
I would like to support the Kubeflow [TrainJob](https://github.com/kubeflow/training-operator/tree/69094e16309382d929606f8c5ce9a9d8c00308b1/docs/proposals/2170-kubeflow-training-v2), which is Kubeflow TrainingOperator v2.

The new TrainJob AP relies on the JobSet API, but we should not enqueue the entire TrainJob or JobSet since the TrainJob uses the JobSet `DependOn` feature, which has some Kueue collaboration problems, as we mentioned [there](https://github.com/andreyvelich/jobset/blob/ea622a6d6ba8cfd876a7d662da2e141c977ecddc/keps/672-serial-job-execution/README.md#risks-and-mitigations)

Hence, in the short term, the TrainJob queueing should rely on the Kueue batch/v1 Job integration to mitigate and avoid the problems once the TrainJob is submitted. This concept is similar to Deployment integration, which depends on the Pod integration. 

Or we might be able to extend the JobSet integration so that the JobSet with dependsOn will be enqueued by batch/v1 Job integration to avoid the long-term unused computing resources lock.

**Why is this needed**:
Currently, we are supporting the Kubeflow v1 APIs, but the v1 APIs will be stopped support within the next week.
We will probably completely stop the v1 support in the next 2 or 3 Training Operator releases.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

**Comment by tenzen-y** [comment]
cc: @andreyvelich @mimowo @kannon92

**Comment by kannon92** [comment]
> Hence, in the short term, the TrainJob queueing should rely on the Kueue batch/v1 Job integration to mitigate and avoid the problems once the TrainJob is submitted. This concept is similar to Deployment integration, which depends on the Pod integration.

TrainJob -> JobSet -> (Job A + JobB)

I think you'd want to explore integration with JobSet before you started with jobs.

What exactly is the end state of integration? Do we want DependsOn or are we fine with support via JobSet? 

If we want to support DependsOn I think Kubeflow community should focus on implementing that in JobSet. Whatever we add we will have to support. So "Short Term" features would end up being long term features as we can't really drop support once customers use it.

**Comment by tenzen-y** [comment]
> If we want to support DependsOn I think Kubeflow community should focus on implementing that in JobSet. Whatever we add we will have to support. So "Short Term" features would end up being long term features as we can't really drop support once customers use it.

I did not indicate that we aim to implement `dependsOn` in the Kubeflow side.
The motivation for using batch/v1 Job Kueue integration is https://github.com/andreyvelich/jobset/blob/ea622a6d6ba8cfd876a7d662da2e141c977ecddc/keps/672-serial-job-execution/README.md#risks-and-mitigations.

Even if Kubeflow uses the JobSet `DependsOn`, we should use the batach/v1 Job integration in the short term.
In the long term, we need to consider how we can separately enqueue each ReplicatedJob as we mentioned in the JobSet `DependsOn` KEP.

**Comment by kannon92** [comment]
@andreyvelich @tenzen-y is there any idea of timeline when this can be implemented?

I know of some people that are interested in using the finetuning functionality with Kueue so it would be worth prioritizing this effort if possible.

**Comment by tenzen-y** [comment]
> [@andreyvelich](https://github.com/andreyvelich) [@tenzen-y](https://github.com/tenzen-y) is there any idea of timeline when this can be implemented?
> 
> I know of some people that are interested in using the finetuning functionality with Kueue so it would be worth prioritizing this effort if possible.

Before we start natively supporting TrainJob in Kueue, we want to bump JobSet API version to Beta or GA to avoid multi API versions supporting.
Additionally, users can now enqueue TrainJob via Kueue JobSet integration, as Kueue supports limited integration.

@kannon92 Do you see any lack of Kueue features for TrainJob via JobSet Kueue integration?

**Comment by andreyvelich** [comment]
>Additionally, users can now enqueue TrainJob via Kueue JobSet integration, as Kueue supports limited integration.

@tenzen-y Given that we have `suspend` and `managedBy` API on the TrainJob side, we still require integration with Kueue Job framework, isn't?
Also, TrainJob handles some custom orchestration of Kubernetes resources (e.g. MPI Plugin).

Additionally, TrainJob might introduce different state machine compare to JobSet, which needs to be integrated with Kueue and MultiKueue.

**Comment by tenzen-y** [comment]
> > Additionally, users can now enqueue TrainJob via Kueue JobSet integration, as Kueue supports limited integration.
> 
> [@tenzen-y](https://github.com/tenzen-y) Given that we have `suspend` and `managedBy` API on the TrainJob side, we still require integration with Kueue Job framework, isn't? Also, TrainJob handles some custom orchestration of Kubernetes resources (e.g. MPI Plugin).
> 
> Additionally, TrainJob might introduce different state machine compare to JobSet, which needs to be integrated with Kueue and MultiKueue.

Yes, we are still eager to natively support TrainJob in Kueue since the JobSet integration is not able to MPI TrainJob, as you said. However, before doing that, KF and JobSet community should mature the APIs.

**Comment by andreyvelich** [comment]
Shall we search for contributors to support TrainJob within Kueue after we make the 2.0 release ?
Given that many customers will migrate from Training Operator V1 to V2, I feel like the integration with Kueue is a high priority.

**Comment by tenzen-y** [comment]
As we discussed with @tenzen-y, @andreyvelich, and @mimowo in KubeCon Tokyo, we decided to support TrainJob v1alpha1, which means we will just drop older API versions from Kueue once a more mature API like v1alpha2 / v1beta1 API.

Additionally, the better management of the `dependsOn` feature is out of scope for this initial TrainJob support. We should appropriately handle it in the future Kueue release.

**Comment by amy** [comment]
/assign kaisoz

(Just correlating the issues. I see that @kaisoz has volunteered here: https://github.com/kubernetes-sigs/kueue/issues/5719#issuecomment-3035560865
And this issue is 0.14.0 release candidate.)

**Comment by kannon92** [comment]
With https://github.com/kubernetes-sigs/kueue/pull/6597 merging, I think we can call this done.

/close

**Comment by kannon92** [comment]
Please feel free to reopen if you want to track the follows up here.
